api/lib/domain/services/badge-criteria-service.js

```diff
@@ -70,6 +57,60 @@ function _verifySomePartnerCompetenceResultsMasteryPercentageCriterion(partnerCo
   return !_.isEmpty(filteredPartnerCompetenceResults);
 }
 
+function _removeUntargetedKnowledgeElements(knowledgeElements, targetProfileSkillsIds) {
+  return _.filter(knowledgeElements, (ke) => targetProfileSkillsIds.some((skillId) => skillId === ke.skillId));
+}
+
+function _computePartnerCompetenceResults(badge, targetProfileSkillsIds, targetedKnowledgeElements) {
+  if (_.isEmpty(badge.badgePartnerCompetences)) {
+    return [];
+  }
+
+  const validTargetedPartnerCompetences = _keepValidTargetedPartnerCompetences(badge.badgePartnerCompetences, targetProfileSkillsIds);
+  return _.map(validTargetedPartnerCompetences, (badgePartnerCompetence) => _getTestedCompetenceResults(badgePartnerCompetence, targetedKnowledgeElements));
+}
+
+function _keepValidTargetedPartnerCompetences(badgePartnerCompetences, targetProfileSkillsIds) {
+  const targetedCompetences = _removeUntargetedSkillIdsFromPartnerCompetences(badgePartnerCompetences, targetProfileSkillsIds);
+  return _removePartnerCompetencesWithoutAnyTargetedSkillsLeft(targetedCompetences);
+}
+
+function _removeUntargetedSkillIdsFromPartnerCompetences(badgePartnerCompetences, targetProfileSkillsIds) {
+  return _.map(badgePartnerCompetences, (badgePartnerCompetence) => {
+    badgePartnerCompetence.skillIds = _.intersection(badgePartnerCompetence.skillIds, targetProfileSkillsIds);
+    return badgePartnerCompetence;
+  });
+}
+
+function _removePartnerCompetencesWithoutAnyTargetedSkillsLeft(badgePartnerCompetences) {
+  return _.filter(badgePartnerCompetences, (badgePartnerCompetence) => !_.isEmpty(badgePartnerCompetence.skillIds));
+}
+
+function _computeMasteryPercentage({ totalSkillsCount, validatedSkillsCount }) {
+  if (totalSkillsCount !== 0) {
+    return Math.round(validatedSkillsCount * 100 / totalSkillsCount);
+  } else {
+    return 0;
+  }
```

Comment:
suggestion: Éviter les conditions négatives
```suggestion
  if (totalSkillsCount === 0) {
    return 0;
  } else {
    return Math.round(validatedSkillsCount * 100 / totalSkillsCount);
  }
```